keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,525 @@
+
+OBCs reference a storage class. The storage class references the external provisioner, defines a reclaim policy, and specifies object-store specific parameters, such as region, owner secret, and bucket lifecycle polcies.
+For brownfield provisioning, the storage class also specifies the name of the existing bucket so that app developers are not burdened with typically long, random bucket names.
+
+### Moving Parts
+
+Before deploying an object bucket consuming app, the OB and OBC CRDs need to be created with RBAC rules granting _get_, _list_, and _watch_ verbs.
+An object store needs to be created and referenced in the storage classes used by OBCs.
+Depending on the object store, a secret in the object-store's namespace, containing new bucket owner credentials, may need to be created.
+The OBCs needed by the app should exist in the app's namespace.
+
+As is true for dynamic PV provisioning, a bucket provisioner pod needs to be running for each object-store created in the Kubernetes cluster.
+For example, for AWS S3, the developer creates an OBC referencing a storage class which references the S3 store.
+This storage class needs to be created by the admin.
+The S3 provisioner pod watches for OBCs whose storage classes point to the AWS S3 provisioner, while ignoring all other OBCs.
+Likewise, the same cluster can also run the Rook-Ceph RGW provisioner, which also watches OBCs, only handling OBCs that reference storage classes which define ceph-rgw.
+
+An OBC for a new bucket triggers the correct provisioner to create a new bucket and associated artifacts (credentials, policy, etc). 
+The library, in turn, creates a secret and configmap based on information returned by the provisioner.
+The secret contains the credentails needed to access the bucket and the configmap contains bucket endpoint information.
+**Note:** it was mentioned at the 2019 NA Kubecon storage face-to-face that to improve scaling we could decide to move the configmap info into the secret, thus reducing k8s resources required.
+The secret and configmap reside in the same namespace as the OBC.
+The app pod consumes the secret and configmap and thus can reference the bucket.
+
+An OBC for an existing bucket triggers the correct provisioner to grant access to the existing bucket. Like new bucket provisioning, associated bucket related artifacts (credentials, policy, etc) are typically created by the provisioner.
+
+When a _greenfield_ OBC is deleted the associated provisioner is expected to delete the newly provisioned bucket and related object store-specific artifacts.
+The library will delete the secret, configmap and OB.
+A deleted _brownfield_ OBC is similar except that provisioners typically will not delete the physical bucket. Instead, bucket access is revoked and related artifacts are expected to be cleaned up.
+Again, the library will delete the secret, configmap and OB.
+
+### API
```

Comment:
"API" is used interchangeably up to this point to reference both the k8s API (CRDs) and the library's interface.  How about defining terms for each at the start ("API" == CRDs, "Interface" == library methods)